### PR TITLE
Fcb erase value support

### LINKED ISSUE #28693: FCB support for non-0xFF flash erase values
**Is your enhancement proposal related to a problem? Please describe.**
Currently FCB is able to work with flash devices that erase to 0xFF values.
Since it's implementation new devices have appeared that use non-0xFF erase values (0x00), which is not supported
by the FCB as it relies on 0xFF as an erased/empty value in some cases.

**Describe the solution you'd like**
Make FCB flash erase value independent.

**Describe alternatives you've considered**
There is none as long as FCB is to be used.

**Additional context**
The flash API now supports  `flash_get_parameters` that returns `flash_parameters` structure that has `erase_value` member, which is initialized by driver with the erase value.

### PR DESCRIPTION
**Pull request by de-nordic**
Adds support to non-0xff erasable flash devices to FCB.

Resolves #28693

### CODE REVIEW

**Comment by nashif** [comment]
@nvlsianpu please review

**Comment by nvlsianpu** [comment]
sure

**Review by nvlsianpu — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@erwango This might be used now on some STM32L1 SoCs which has 0x00 as flash erased state.

**Comment by utzig** [comment]
> @erwango This might be used now on some STM32L1 SoCs which has 0x00 as flash erased state.

I was gonna try it a few weeks ago but seems like the stm32 flash driver does not support L1 as of yet.

**Comment by de-nordic** [comment]
> I was gonna try it a few weeks ago but seems like the stm32 flash driver does not support L1 as of yet.

Thanks @utzig !

**Comment by nvlsianpu** [comment]
@erwango @utzig ping

**Comment by erwango** [comment]
Thanks for pinging me. Though, I don't have ways to test this right now and I'm not sure my review has much added value.
What I would suggest though, is to move erase-value prop value definition from zephyr,sim-flash.yaml to flash-controller.yaml so that it could be used widely. Not sure this is the topic of this PR though.

**Comment by nvlsianpu** [comment]
@erwango ^^ PR is no related to where erase-value property is defined.

File: subsys/fs/fcb/fcb.c [hunk 1]
```diff
@@ -194,35 +205,54 @@ fcb_is_empty(struct fcb *fcb)
...
+fcb_put_len(const struct fcb *fcb, uint8_t *buf, uint16_t len)
 {
 	if (len < 0x80) {
-		buf[0] = len;
+		buf[0] = len ^ ~fcb->f_erase_value;
 		return 1;
 	} else if (len < FCB_MAX_LEN) {
-		buf[0] = (len & 0x7f) | 0x80;
-		buf[1] = len >> 7;
+		buf[0] = ((len & 0x7f) | 0x80) ^ ~fcb->f_erase_value;
```
Comment (utzig) [inline subsys/fs/fcb/fcb.c:233]: This is a nitpick, but can `((len & 0x7f) | 0x80)` ever result in something that is not equal to just `(len | 0x80)` when you're storing it in a `uint8_t`?

**Review by utzig — APPROVED** [review]
LGTM, left some nitpicks.

File: subsys/fs/fcb/fcb.c [hunk 2]
```diff
@@ -12,6 +12,11 @@
 #include "fcb_priv.h"
 #include "string.h"
 #include <errno.h>
+#include <device.h>
+#include <drivers/flash.h>
+
+#define MK32(val) ((((uint32_t)val) << 24) | ((val) << 16) | \
+			      ((val) << 8) | (val))
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply subsys/fs/fcb/fcb.c]: Fixed.

File: subsys/fs/fcb/fcb_priv.h [hunk 3]
```diff
@@ -17,15 +17,34 @@ extern "C" {
...
+ *
+ * @param fcb pointer to initialized fcb structure
+ *
+ * @return uin32_t formatted magic value
+ */
+static inline uint32_t fcb_flash_magic(const struct fcb *fcb)
+{
+	const uint8_t ev = fcb->f_erase_value;
+
+	return (fcb->f_magic ^ ~((ev << 24) | (ev << 16) | (ev << 8) | ev));
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply subsys/fs/fcb/fcb_priv.h]: Fixed.

File: subsys/fs/fcb/fcb.c (hunk 1 above)
Comment (de-nordic) [reply subsys/fs/fcb/fcb.c:230]: I doubt.

**Review by utzig — APPROVED** [review]

**Comment by nvlsianpu** [comment]
cc @carlescufi This is ready.

**Comment by carlescufi** [comment]
@de-nordic can you rebase this? seems like the doc build error was solved in master

**Pull request merged**